piilo/engines/anonymizer.py

```diff
@@ -1,60 +1,108 @@
+        Dict[PII_TYPE[PII_VALUE]] = [REPLACEMENT]
 
-    def get_gender(self, name: HumanName) -> str:
-        """Return the most frequent gender code for the provided name's
-        first name, or None if a match cannot be found.
-        """
-        result = self.first_names.get(name.first)
-        return self._max(result.get("gender", None)) if result else None
+    Args:
+        AnonymizerEngine (presidio_anonymizer.AnonymizerEngine):
+        A Presidio anonymizer engine
+    """
 
-    def get_country(self, name: HumanName) -> str:
-        """Return the most frequent country code for the provided name's
-        last name, or None if a match cannot be found.
-        """
-        result = self.last_names.get(name.last)
-        return self._max(result.get("country", None)) if result else None
+    names_df_path = "data/ascii_names.parquet"
+    obfuscation_map_path = None
+    date_digits = re.compile(r"(?:\d{4}|\d{1,2})")
 
+    def __init__(self, remember_replacements: Literal["all", "document"] = "all"):
+        """
+        Args:
+            remember_replacements (Literal["all", "document"], default="all"):
+            Surrogate PII will be replaced with the same value every time it is
+            seen within the configured scope.
 
-class SurrogateAnonymizer(AnonymizerEngine):
-    """A wrapper around the presidio_anonymizer.AnonymizerEngine class."""
+            "all" will remember all replacements made during the session.
+            This may be useful for related documents, such as a collection of
+            discussion board posts.
 
-    def __init__(self):
+            "document" will remember replacements made during the current document.
+            This is optimal for unrelated documents, such as a collection of essays.
+        """
         super().__init__()
-        self.names_db = NameDatabase()
-        self.names_df = pd.read_parquet(name_table)
+        self.fake = Faker()
+
+        self.name_getter = NameGetter()
+        self.names_df = pd.read_parquet(self.names_df_path)
+
+        # Configurable mappings for PII --> Surrogate for PII labeled as "OTHER".
+        # We may decide not to use this, or to develop an interface for
+        # generating these mappings "on-the-fly" in an interactive obfuscation mode.
+        if self.obfuscation_map_path:
+            with open(self.obfuscation_map_path, "r") as f:
+                self.obfuscation_map = json.load(f)
+        else:
+            self.obfuscation_map = {}
+
+        # TODO: To enable the shuffle_obfuscate method, we need a complete list
+        # of PII values for each type. This will need to be generated from input data.
+        # Form should be {"PII_TYPE": [PII_VALUES]}
+        self.all_pii = {}
+
+        # keep track of PII we have seen and its surrogate replacement
+        self.remember = remember_replacements
+        self.seen_pii = defaultdict(dict)
+
+    def set_seeds(self, text: str) -> None:
```

Comment:
Do we want to keep the same seed for the same document in production (do we want the same document to produce same anonymized names every time)?

Reply:
Good question! I think we might want to make it a configuration option. We shouldn't do this without providing an option to turn it off. Low priority though. We could just remove this entirely if we run out of time.

This is totally accidental, but I think it is actually secure since the hash is built from the raw document. Even if someone were to examine the PIILO source code, they could not deduce the raw document's hash from the obfuscated document. As a result, we don't need to worry about somebody figuring out which PII was obfuscated based on this non-random process (and finding out which names/PII have been obfuscated would defeat the privacy benefits of hiding-in-plain-sight).